### PR TITLE
codec: support half-close in HTTP

### PR DESCRIPTION
**Pull request by kyessenov**
Signed-off-by: Kuat Yessenov <kuat@google.com>

Commit Message: Hold ActiveRequest destruction until a full close instead of doing immediately on response decode.
Additional Description: This is needed to support half closed tcp_proxy CONNECT streams. Closing immediately on response triggers TCP connection close on the draining upstream which breaks tcp_proxy continuing to send data. Fixes https://github.com/envoyproxy/envoy/issues/22748. See H2 state transition diagram here: https://httpwg.org/specs/rfc9113.html#StreamStates.

HTTP/1.1 is particularly difficult to change because message completion is triggered by the parser. This PR keeps the behavior unchanged for HTTP/1.1 codec.

QUIC requires adding STOP_SENDING support on client side, otherwise, the underlying stream is destroyed before the callbacks kick in. The test in protocol integration is updated to ensure that test processes client STOP_SENDING before closing the connection.

Risk Level: medium / high
Testing: manually verified the crash is fixed, reproducing test case added
Docs Changes: none
Release Notes: yes
Runtime feature: envoy.reloadable_features.http_response_half_close

### CODE REVIEW

File: envoy/http/codec.h
Comment (alyssawilk) [inline envoy/http/codec.h:574]: I think this is a bit confusing due to encode/decode meaning separate things for streams vs codecs.

maybe qualify a bit more:?

True if the codec should await for the request end stream to be encoded to the codec in addition to the response end stream being decoded from the codec before the stream can be closed.

File: source/common/http/codec_client.cc
Comment (alyssawilk) [inline source/common/http/codec_client.cc:75]: If this is just based on protocol, can't we switch on CodecType here and skip the new interface entirely?

I think we should also comment (here or if you stick with the interface, there) why we skip this for HTTP/1.1

File: source/common/http/codec_wrappers.h
Comment (alyssawilk) [inline source/common/http/codec_wrappers.h:127]: can we do the rename in a follow up? easier to see the diffs if they're only functonal

File: source/common/http/codec_client.h
Comment (alyssawilk) [inline source/common/http/codec_client.h:226]: wait how did we go from awaiting to deciding the encode was complete?
Ah on reflection I assume this is for simplicity since we're "not tracking" but it's going to confuse some deugging some day so let's not shortcut here :-)

**Review by alyssawilk — COMMENTED** [review]
yeah looks much simpler thanks.  have a few more clean up requests.  definitely want an integration test, which I think you can work on in parallel since that'll be reusable regardless of how we fix
/wait

File: source/common/http/codec_client.cc
Comment (kyessenov) [reply source/common/http/codec_client.cc:75]: OK, we can do that to keep it simpler.

File: envoy/http/codec.h
Comment (kyessenov) [reply envoy/http/codec.h:574]: Ack, will update in the nest commit (resolving quic sanitizer issue).

File: source/common/http/codec_client.h
Comment (kyessenov) [reply source/common/http/codec_client.h:226]: Ack, will improve readability.

File: source/common/quic/envoy_quic_client_stream.cc
Comment (alyssawilk) [inline source/common/quic/envoy_quic_client_stream.cc:231]: cc @danzh2010 can you look at this one

**Review by alyssawilk — COMMENTED** [review]
LGTM modulo Dan or Ryan taking a look at the HTTP bits.
Matt would you be up for second pass?

File: source/common/quic/envoy_quic_client_stream.cc
Comment (danzh2010) [inline source/common/quic/envoy_quic_client_stream.cc:233]: Plz add `stats_.rx_reset_.inc();`

Comment (kyessenov) [reply source/common/quic/envoy_quic_client_stream.cc:233]: Added, thanks.

File: source/common/quic/envoy_quic_client_stream.cc
Comment (kyessenov) [reply source/common/quic/envoy_quic_client_stream.cc]: Good point. I think it was working before because the server would only STOP_SENDING after closing, but that cannot be assumed in general.

I'm not sure about OnStreamReset(). What would be broken if we leave it as-is?

File: source/common/quic/envoy_quic_client_stream.cc
Comment (kyessenov) [reply source/common/quic/envoy_quic_client_stream.cc]: Thanks, got confused by client/server/read/write/decode/encode.

File: source/common/quic/envoy_quic_client_stream.cc
Comment (kyessenov) [reply source/common/quic/envoy_quic_client_stream.cc:233]: BTW adding here seems to break multiplexed_upstream_integration_test so I moved it closer to actual callback. It seems we shouldn't increment in the normal case after local stream end.

Comment (danzh2010) [reply source/common/quic/envoy_quic_client_stream.cc:233]: That makes sense, the stats increments when the stream receives both RESET and STOP_SENDING.

File: source/common/quic/envoy_quic_client_stream.cc
Comment (danzh2010) [reply source/common/quic/envoy_quic_client_stream.cc]: > I'm not sure about OnStreamReset(). What would be broken if we leave it as-is?

You need to check write_side_closed() before calling runResetCallbacks(), because OnStreamReset() only close read side.

Comment (kyessenov) [reply source/common/quic/envoy_quic_client_stream.cc]: Updated to mirror server side. Also updated a test that was incorrect.

Comment (danzh2010) [reply source/common/quic/envoy_quic_client_stream.cc]: Thanks for the additional tests!

Comment (danzh2010) [reply source/common/quic/envoy_quic_client_stream.cc]: Does the IntegrationStreamDecoder lifetime issue still exist now?

Comment (kyessenov) [reply source/common/quic/envoy_quic_client_stream.cc]: Let me try to push a commit and see? I can't reproduce the sanitizer issue locally for some reason, only on CI.

Comment (kyessenov) [reply source/common/quic/envoy_quic_client_stream.cc]: Yes, the lifetime issue still exists, see this log https://dev.azure.com/cncf/envoy/_build/results?buildId=115672&view=logs&jobId=4e2e4662-2d21-5726-b4bc-9ff5cd6a179b&j=4e2e4662-2d21-5726-b4bc-9ff5cd6a179b&t=a4553756-733d-5e3b-e02d-5aef5f5e9bee. I'll revert the test commit.

Comment (danzh2010) [reply source/common/quic/envoy_quic_client_stream.cc]: Sorry, the link doesn't load the test result for me. What's the msan error?

Comment (kyessenov) [reply source/common/quic/envoy_quic_client_stream.cc]: Argh, the log disappeared for me, too :( It's msan violation, access to deleted IntegrationStreamDecoder from Reset callback.

Comment (kyessenov) [reply source/common/quic/envoy_quic_client_stream.cc]: The test is `FaultyFilterWithConnect`.

Comment (danzh2010) [reply source/common/quic/envoy_quic_client_stream.cc]: It's a msan failure in Linux, can you repro the test locally? It'll great if you can explain a bit more about what the life cycle difference between HTTP/3 and HTTP/2 that caused this failure. It's unclear to me if this is a test setup issue or a newly introduced bug in quic client stream objects life cycle.

**Review by mattklein123 — DISMISSED** [review]
At a high level this looks good to me. Any way you slice it this is a very risky change so in a perfect world we would get someone to smoke test it. I'm also on the fence whether this should be runtime guarded given the risk level? @alyssawilk? Testing looks extensive though. Thank you! I will defer to others for final review/merge.

File: source/common/quic/envoy_quic_client_stream.cc
Comment (kyessenov) [reply source/common/quic/envoy_quic_client_stream.cc]: Let me try to undo test changes and instead just close the request. I think that is more reasonable.

Comment (kyessenov) [reply source/common/quic/envoy_quic_client_stream.cc]: Sample error https://gist.github.com/kyessenov/016643c55bf2c34863954c76b9f805dd

**Comment by kyessenov** [comment]
Current status: debugging one integration test for QUIC failing due to a race with STOP_SENDING receive and client close.

**Comment by kyessenov** [comment]
@mattklein123 I can add a default-true runtime override. I think that should mitigate the risk. We did notice this PR improved our CI integration testing for a mesh where all connections are CONNECTs and server would often abruptly close itself.

File: source/common/quic/envoy_quic_client_stream.cc
Comment (kyessenov) [reply source/common/quic/envoy_quic_client_stream.cc]: The root cause was determined to be the early test termination before client had a chance to read STOP_SENDING and fully reset before the connection closure. Thank you @danzh2010 for assistance.

Comment (danzh2010) [reply source/common/quic/envoy_quic_client_stream.cc]: You're welcome! Side note: can https://github.com/envoyproxy/envoy/pull/22802/commits/b18d968bffbc56024ac775f109b9d3d750daa976 be applied forward now?

Comment (kyessenov) [reply source/common/quic/envoy_quic_client_stream.cc]: I actually removed all the changes to the test framework, so that commit is no longer necessary. The PR is much smaller now.

**Review by danzh2010 — COMMENTED** [review]
LGTM, quic changes

**Comment by kyessenov** [comment]
Waiting for final review by @alyssawilk .

**Review by alyssawilk — APPROVED** [review]

**Pull request merged**
